File: 0904-leaf-similar-trees/0904-leaf-similar-trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        leaves1=[]
        leaves2=[]
        def dfs1(node):
            if node is None:
                return
            if node.left is None and node.right is None:
                leaves1.append(node.val)
            dfs1(node.left)
            dfs1(node.right)
        dfs1(root1)
        def dfs2(node):
            if node is None:
                return
            if node.left is None and node.right is None:
                leaves2.append(node.val)
            dfs2(node.left)
            dfs2(node.right)
        dfs2(root2)
        if leaves1==leaves2:
            return True

File: 0904-leaf-similar-trees/0904-leaf-similar-trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void Totalleaves(TreeNode* node, vector<int>& leaves) {
        if(node==nullptr){
            return;
        }
        if(node->left==nullptr && node->right==nullptr){
            leaves.push_back(node->val);
        }
        Totalleaves(node->left,leaves);
        Totalleaves(node->right,leaves);
    }
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        vector<int>leaves1,leaves2;
        Totalleaves(root1,leaves1);
        Totalleaves(root2,leaves2);
        
        return leaves1==leaves2;
    }
};